src/main/java/com/synopsys/integration/blackduck/bdio2/Bdio2FileUploadService.java

```diff
@@ -85,7 +85,7 @@ private Bdio2UploadResult uploadFiles(List<BdioFileContent> bdioFiles, @Nullable
 
         WaitIntervalTracker waitIntervalTracker = WaitIntervalTrackerFactory.createConstant(timeout, BD_WAIT_AND_RETRY_INTERVAL);
         ResilientJobConfig jobConfig = new ResilientJobConfig(logger, System.currentTimeMillis(), waitIntervalTracker);
-        Bdio2UploadJob bdio2UploadJob = new Bdio2UploadJob(bdio2RetryAwareStreamUploader, header, remainingFiles, editor, count, shouldUploadEntries, shouldFinishUpload);
+        Bdio2UploadJob bdio2UploadJob = new Bdio2UploadJob(bdio2RetryAwareStreamUploader, header, remainingFiles, editor, count, shouldUploadEntries, shouldFinishUpload, clientStartTime, timeout);
```

Comment:
Do you know how the BD_WAIT_AND_RETRY_INTERVAL in line 86 affects our new retry logic? Not entirely familiar with how retries worked before the ask in IDETECT4068

Reply:
No worries, this code is fairly complex. If the below doesn't help feel free to reach out and we can go through it in detail.

The wait interval here is tied to the ResilientJob framework that Bdio2UploadJob is a part of (if you look at it, it implements ResilentJob). That is going to do a Fibonacci based backoff which is essentially what we want to happen when we don't get a retry-after header from BlackDuck. Bdio2UploadJob uses it when it starts a scan, when it appends to a scan, and when it finishes a scan.

When starting a scan, and only when starting a scan, we want to see if we get a retry-after header from BlackDuck that tells us exactly how long to wait. This is why the changes I've made are largely confined to Bdio2RetryAwareStreamUploader.start. If we get that retry-after we sleep on it and then try to start the scan again. If things are successful or not (and we don't have this header) we return the response. The ResilentJob framework will then complete the scan for us, if it was successful, or do a Fibonacci backoff if it was not.